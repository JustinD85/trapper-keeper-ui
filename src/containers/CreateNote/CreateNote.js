import React from "react";
import { connect } from "react-redux";
import { addNewNote, updateNote } from "../../actions";
import { postNote, putNote } from "../../thunks";
import PropTypes from "prop-types";
import uuid from "uuid/v4";
import { withRouter } from "react-router-dom";
import { Dialog, DialogContent, DialogTitle, Typography, List, ListItem, ListItemText, Button, Checkbox, IconButton, Slide} from '@material-ui/core';
import { Delete } from '@material-ui/icons'

export class CreateNote extends React.Component {
  constructor(props) {
    super(props);
    this.state = {
      title: this.props.title || "",
      noteItems: this.props.noteItems || [],
      currentFocus: null
    };
  }

  handleChangeTitle = event => {
    this.setState({ title: event.target.value, currentFocus: null });
  };

  makeCopy = element => JSON.parse(JSON.stringify(element))
  
  handleChangeNoteItems = event => {
    const noteItemsCopy = this.makeCopy(this.state.noteItems);
    const { value: newText } = event.target;
    const { id } = event.target.closest('label')
    const matchedNoteItem = noteItemsCopy.find(note => note.id === id);

    if (matchedNoteItem) {
      matchedNoteItem.text = newText;     
    } else {
      const newListItem = { id, text: newText, isCompleted: false };
      noteItemsCopy.push(newListItem);
    }

    this.setState({
      noteItems: noteItemsCopy,
      currentFocus: id
    });
  };

  handleToggleIsComplete = event => {
    const noteItemsCopy = this.makeCopy(this.state.noteItems);
    const { id } = event.target.closest('label');
    const matchedNoteItem = noteItemsCopy.find(note => note.id === id);

    if(matchedNoteItem) {
      matchedNoteItem.isCompleted = !matchedNoteItem.isCompleted
    } else {
      const newListItem = {id, text: '', isCompleted: event.target.checked}
      noteItemsCopy.push(newListItem);
    }

    this.setState({
      noteItems: noteItemsCopy,
      currentFocus: id
    })
  };

  handleItemDelete = (event) => {
    const noteItemsCopy = this.makeCopy(this.state.noteItems);
    const { id } = event.target.closest('label');
    const noteItemIndex = noteItemsCopy.findIndex(note => note.id === id);
    noteItemsCopy.splice(noteItemIndex, 1);
    this.setState({ 
      noteItems: noteItemsCopy
    });
  }

  getListItems() {
    const { noteItems } = this.state;
    let currentList = noteItems.map(item => {
      let jsxNoteItem = (
        <ListItem key={uuid()}>
          <label id={item.id}>
            <Checkbox onChange={this.handleToggleIsComplete} checked={item.isCompleted}/>
            <input
              key={item.id}
              autoFocus={item.id === this.state.currentFocus}
              onChange={this.handleChangeNoteItems}
              value={item.text}
            />
            <IconButton onClick={this.handleItemDelete}>
            <Delete />
            {/* <img src="./images/trash-can-outline.svg"/> */}
            </IconButton>
          </label>
        </ListItem>
      );
      return jsxNoteItem;
    });

    currentList.push(
      <ListItem key={uuid()}>
        <label id={uuid()}>
          <input key={uuid()} onChange={this.handleChangeNoteItems} />
        </label>
      </ListItem>
    );
    return currentList;
  }

  handleSubmit = event => {
    event.preventDefault();
    const { title, noteItems } = this.state;
    const { id: noteId, putNote, postNote } = this.props;
    if (noteId) {
      putNote({ title, noteItems, id: noteId });
    } else {
      postNote({ title, noteItems, id: uuid() });
    }
    this.props.history.push("/");
  };

  render() {
    const { title } = this.state;
    let isOpen = this.props.location.pathname.includes('note')
    return (
      <Dialog onClose={() => this.props.history.push('/')} open={isOpen} transitionDuration={1000} TransitionComponent={(props) => <Slide direction='up' {...props}/>}>
      <DialogTitle>
        <input value={title} onChange={this.handleChangeTitle} placeholder='Add a title'/>
      </DialogTitle>
      <DialogContent>
          <form onSubmit={this.handleSubmit}>
            <List>{this.getListItems()}</List>
            <Button type='submit'>Submit</Button>
          </form>
      </DialogContent>
        </Dialog> 
    );
  }
}

export const mapDispatchToProps = dispatch => ({
  addNewNote: newNote => dispatch(addNewNote(newNote)),
  postNote: newNote => dispatch(postNote(newNote)),
  putNote: updatedNote => dispatch(putNote(updatedNote)),
  updateNote: updatedNote => dispatch(updateNote(updatedNote))
});

export default withRouter(
  connect(
    null,
    mapDispatchToProps
  )(CreateNote)
);

CreateNote.propTypes = {
  addNewNote: PropTypes.func,
  postNote: PropTypes.func,
  putNote: PropTypes.func,
  updateNote: PropTypes.func
};
