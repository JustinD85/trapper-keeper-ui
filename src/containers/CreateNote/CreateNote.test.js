import React from 'react';
import { shallow } from 'enzyme';
import { CreateNote } from './CreateNote';
import uuid from "uuid/v4";

jest.mock('uuid/v4');
let findMatchingNoteItemMock = jest.fn([{title: 'title', id: 1}]);
let mockEvent = { target: { value:'newTitle', closest: () => ({ id: 1 }) }};
let noteItemsMock = [{ id: 1, text: 'test', isCompleted: false }];

describe('CreateNote', () => {
  let wrapper;
  beforeEach(() => {
    wrapper = shallow(<CreateNote location={{ pathname: 'a' }}
                                  findMatchingNoteItem={findMatchingNoteItemMock}
    />);
  });

  it("should render with no error and match snapshot", () => {
    expect(wrapper).toMatchSnapshot();
  });

  it('should have the proper default state', () => {
    expect(wrapper.state()).toEqual({
      title: '',
      noteItems: [],
      currentFocus: null
    });
  });

  it('should change the title and currentFocus in state when handleChangeTitle is called', () => {
    expect(wrapper.state('title')).toEqual('');
    expect(wrapper.state('currentFocus')).toEqual(null);
    wrapper.instance().handleChangeTitle(mockEvent);
    expect(wrapper.state('title')).toEqual('newTitle');
    expect(wrapper.state('currentFocus')).toEqual(null);
  });

  it.skip('should change the matched item text when handleChangeNoteItems is called if the item already exists', () => {
    wrapper.setState({ noteItems: [{ text: 'mockTitle', id: 1, isCompleted: true }] });
    wrapper.instance().handleChangeNoteItems(mockEvent);
    expect('matchedNoteItem.text').toEqual('mockTitle');
  });

  it.skip('should add a new item when handleChangeNoteItems is called if the item does not exist', () => {});

  it.skip('should update the state of noteItems and currentFocus when handleItemDelete is called', () => {
    const updateNoteItems = jest.fn();
    wrapper.setState({ noteItems: [{ text: 'mockTitle', id: 1, isCompleted: true }] });
    wrapper.instance().handleItemDelete(mockEvent);
  });

  it('should update the state of noteItems and currentFocus when updateNoteItems is called', () => {
    expect(wrapper.state('noteItems')).toEqual([]);
    expect(wrapper.state('currentFocus')).toEqual(null);
    wrapper.instance().updateNoteItems(noteItemsMock, 1);
    expect(wrapper.state('noteItems')).toEqual(noteItemsMock);
    expect(wrapper.state('currentFocus')).toEqual(1);
  });

  it('should call getListItems when renderListItems is called', () => {
    const getListItems = jest.fn();
    wrapper.instance().renderListItems();
    expect(getListItems).toBeCalled();
  });


}); 