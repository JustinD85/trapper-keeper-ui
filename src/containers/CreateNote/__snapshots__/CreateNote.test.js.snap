// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CreateNote should render with no error and match snapshot 1`] = `
<WithStyles(Dialog)
  onClose={[Function]}
  open={false}
  transitionDuration={1000}
>
  <WithStyles(DialogTitle)>
    <input
      onChange={[Function]}
      placeholder="Add a title"
      value=""
    />
  </WithStyles(DialogTitle)>
  <WithStyles(DialogContent)>
    <form
      onSubmit={[Function]}
    >
      <WithStyles(List)>
        <WithStyles(ListItem)>
          <label>
            <input
              onChange={[Function]}
            />
          </label>
        </WithStyles(ListItem)>
      </WithStyles(List)>
      <WithStyles(Button)
        type="submit"
      >
        Submit
      </WithStyles(Button)>
    </form>
  </WithStyles(DialogContent)>
</WithStyles(Dialog)>
`;
